cni-plugin/deployment/scripts/install-cni.sh

```diff
@@ -55,94 +55,133 @@ CONTAINER_CNI_BIN_DIR=${CONTAINER_CNI_BIN_DIR:-/opt/cni/bin}
 # Default to the first file following a find | sort since the Kubernetes CNI runtime is going
 # to look for the lexicographically first file. If the directory is empty, then use a name
 # of our choosing.
-CNI_CONF_PATH=${CNI_CONF_PATH:-$(find "${CONTAINER_MOUNT_PREFIX}${DEST_CNI_NET_DIR}" -maxdepth 1 -type f \( -iname '*conflist' -o -iname '*conf' \) | sort | head -n 1)}
-CNI_CONF_PATH=${CNI_CONF_PATH:-"${CONTAINER_MOUNT_PREFIX}${DEST_CNI_NET_DIR}/01-linkerd-cni.conf"}
-
+HOST_CNI_NET="${CONTAINER_MOUNT_PREFIX}${DEST_CNI_NET_DIR}"
+DEFAULT_CNI_CONF_PATH="${HOST_CNI_NET}/01-linkerd-cni.conf"
+CNI_CONF_PATH=${CNI_CONF_PATH:-$(find "${HOST_CNI_NET}" -maxdepth 1 -type f \( -iname '*conflist' -o -iname '*conf' \) | sort | head -n 1)}
+CNI_CONF_PATH=${CNI_CONF_PATH:-"${DEFAULT_CNI_CONF_PATH}"}
 KUBECONFIG_FILE_NAME=${KUBECONFIG_FILE_NAME:-ZZZ-linkerd-cni-kubeconfig}
 
+############################
+### Function definitions ###
+############################
+
+# Cleanup will remove any installed configuration from the host If there are any
+# *conflist files, then linkerd-cni configuration parameters will be removed
+# from them; otherwise, if linkerd-cni is the only plugin, the configuration
+# file will be removed.
 cleanup() {
+  # First, kill 'inotifywait' so we don't process any DELETE/CREATE events
+  if [ -n "$(pgrep inotifywait)" ]; then
```

Comment:
is it possible to save the pid of inotifywait using `$!` after launching it in the background so that we have it instead of needing to look it up with pgrep?

Reply:
At first I thought it would be, but now I'm not so sure. Here's what I have tried and the results:

Before commit https://github.com/linkerd/linkerd2/pull/8299/commits/6e88da9947197eb49298d1ccf63c269e915cd3e7 we were spawning `monitor()` in the background and then saving the pid using `$!` as suggested. The problem is that when we sent `SIGKILL` or `SIGTERM` to the script, we would send it to the bash process and not to `inotifywait`. Even though `SIGKILL` (as opposed to `SIGTERM`) should instantly terminate the process, we still processed _one more_ file change.

```
SIGTERM received, simply exiting...
Sending SIGKILL to inotifywait
Removing linkerd-cni artifacts.
/linkerd/install-cni.sh: line 73:    35 Killed                  monitor
Cleaning up /host/etc/cni/net.d/01-linkerd-cni.conf
Removing linkerd-cni kubeconfig: /host/etc/cni/net.d/ZZZ-linkerd-cni-kubeconfig
Detected change in /host/etc/cni/net.d/: DELETE 01-linkerd-cni.conf
Removing linkerd-cni binary: /host/opt/cni/bin/linkerd-cni
```

Not really an error, nothing was added, but it made it weird in the logs. Here's the process tree, along with a log line that shows the pid that we're recording through `$!`

```
root@k3d-cni-server-0:/linkerd# ps -o pid,comm=Command -x -f
    PID Command
     39 bash
     46  \_ ps
      1 bash
     35 bash
     37  \_ inotifywait
     38  \_ bash
     36 sleep
---
# logs
Created CNI config /host/etc/cni/net.d/01-linkerd-cni.conf
WATCH_PID: 35
Setting up watches.
Watches established.
```

Through `pgrep` we send the term directly to `inotifywait` which seems to kill it. I tried one more thing now to see if it would work: instead of spawning the watch in a script, spawn it as a command:

```sh
inotifywait -m "${HOST_CNI_NET}" -e create,delete |
  while read -r directory action filename; do
    if [[ "$filename" =~ .*.(conflist|conf)$ ]]; then
      echo "Detected change in $directory: $action $filename"
      sync "$filename" "$action" "$cni_conf_sha"
      # When file exists (i.e we didn't deal with a DELETE ev)
      # then calculate its sha to be used the next turn.
      if [ -e "$directory/$filename" ]; then
        cni_conf_sha="$(sha256sum "$directory/$filename" | awk '{print $1}')"
      fi
    fi
  done &

WATCH_PID=$!
```

Problem here is that we're making use of the prev sha which is not bound to a variable. We could get around it by probably turning it into an env variable, but that makes it a bit trickier (see ver's feedback). Also, `$!` gives us a weird process number... we can't place variables before the while/read loop either because we have to pipe the event from inotify. I love bash 🙃

```
/linkerd/install-cni.sh: line 314: cni_conf_sha: unbound variable
WATCH_PID: 12
root@k3d-cni-server-0:/linkerd# ps -o pid,comm=Command -x -f
    PID Command
     38 bash
     44  \_ ps
      1 bash
     11 inotifywait
     37 sleep
```

Sorry for the long post but one more thing that might be helpful to figure out a better way here:

```
If bash is waiting for a command to complete and receives a signal for which a trap has been set, the trap will not be executed until the command completes. When bash is waiting for an asynchronous command via the wait builtin, the reception of a signal for which a trap has been set will cause the wait builtin to return immediately with an exit status greater than 128, immediately after which the trap is executed.
```
([ref](https://linux.die.net/man/1/bash) -- last paragraph of "Signals" section)

I think this is why it's so difficult to solve this; `inotify` is a command, `read` is a built-in (as well as any other while loop), but this in effect means `inotify` has to complete, which is why I'm guessing we're always processing one more event when we kill the monitor script?